hlsl-headers/include/vk/khr/cooperative_matrix.impl

```diff
@@ -0,0 +1,216 @@
+          internal::CooperativeMatrixLayout::RowMajorKHR);                     \
+    }                                                                          \
+    void StoreColumnMajor(RWStructuredBuffer<ComponentType> data,              \
+                          uint32_t index) {                                    \
+      internal::CooperativeMatrixStoreKHR(                                     \
+          data[index], _matrix,                                                \
+          internal::CooperativeMatrixLayout::ColumnMajorKHR);                  \
+    }                                                                          \
+                                                                               \
+    template <class BufferType>                                                \
+    static CooperativeMatrix LoadRowMajor(BufferType buffer, uint32_t index) { \
+      CooperativeMatrix result;                                                \
+      result._matrix = internal::CooperativeMatrixLoadKHR<SpirvMatrixType>(    \
+          buffer[index], internal::CooperativeMatrixLayout::RowMajorKHR);      \
+      return result;                                                           \
+    }                                                                          \
+                                                                               \
+    template <class BufferType>                                                \
+    static CooperativeMatrix LoadColumnMajor(BufferType buffer,                \
+                                             uint32_t index) {                 \
+      CooperativeMatrix result;                                                \
+      result._matrix = internal::CooperativeMatrixLoadKHR<SpirvMatrixType>(    \
+          buffer[index], internal::CooperativeMatrixLayout::ColumnMajorKHR);   \
+      return result;                                                           \
+    }                                                                          \
+                                                                               \
+    static uint GetLength() {                                                  \
+      return internal::__builtin_CooperativeMatrixLengthKHR<SpirvMatrixType>();\
+    }                                                                          \
+                                                                               \
+    static const bool isSignedInteger = SIGNED_INTEGER_TYPE;                   \
+    SpirvMatrixType _matrix;                                                   \
+  };
+
+COOPERATIVE_MATRIX(half, spv::FNegate, spv::FAdd, spv::FSub, false);
+COOPERATIVE_MATRIX(float, spv::FNegate, spv::FAdd, spv::FSub, false);
+COOPERATIVE_MATRIX(double, spv::FNegate, spv::FAdd, spv::FSub, false);
+
+// TODO: Need to check for a 16-bit-type enabled
+// COOPERATIVE_MATRIX(int16_t, spv::SNegate, spv::IAdd, spv::ISub, true);
```

Comment:
Need to find a proper solution for 16-bit integers.

Solutions: 

1. We could avoid the macros. We could define two versions of the class. One that uses the float opcodes and another that uses the int opcodes. The uses would be expected to pick the correct one. Then we could define IsSigned by comparing `0-1 > 0`. In Vulkan, this is well defined for both signed and unsigned integers.
2. Use a separate header file for 16-bit instantiations. 

Both of these options make usability a little more difficult.

Reply:
The third option, @keenuts reminded me of, is to have the option define a macro, and we can ifdef the int16_t instantiation.

Reply:
There already was a macro: __HLSL_ENABLE_16_BIT. I used that to guard the template instantiations for int16_t and uint16_t.